### DISCUSSION TITLE
Nested Fragments Result in Type Errors With Fragment Types not Expanded in the Parent Query's Typing

### DISCUSSION THREAD

**Discussion by SimonPringleWallace** [dotansimha/graphql-code-generator#9271] [Q&A] [answered]
Most of the queries in my application are built up of fragments. In many cases these can be rather nested - a query that contains a fragment that's built up of a couple of other fragments, and so on. When using code-gen to generate a type for the parent query I'd expect that the generated type would also include the types for all of the child fragment but that doesn't seem to be true. Whenever I try to access a property that's part of a fragment off of the parent query, I get a TypeScript error that the property isn't found on the type. What am I doing wrong?

This screenshot displays the query, fragment and the error (notice that it's on a property defined within the fragment)
<img width="1427" alt="Screenshot 2023-04-04 at 11 00 19 AM" src="https://user-images.githubusercontent.com/40777383/229836265-295c5885-5974-4b97-abff-529879efd6b8.png">

My config:
```
import type { CodegenConfig } from '@graphql-codegen/cli';

const config: CodegenConfig = {
  schema: 'https://rickandmortyapi.com/graphql',
  documents: 'src/**',
  generates: {
    'src/common/graphql/types/': {
      preset: 'client',
      plugins: [],
    },
  },
};

export default config;
```

A small repo that reproduces the issue: https://github.com/SimonPringleWallace/code-gen-issue/tree/main/src

**Comment by charpeni** [comment, marked as answer]
What's causing that TypeScript error is, in fact, [Fragment Masking](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#fragment-masking), which is enabled by default with `client-preset`.

> Fragment Masking helps express components' data dependencies with GraphQL Fragments.
>
> By doing so, we ensure that the tree of data is properly passed down to the components without "leaking" data. It also allows to colocate the Fragment definitions with their components counterparts:

That's why TypeScript believes `info` doesn't exist on the type, because it is masked. If you want to pursue with fragment masking, the component will be responsible for expressing its data requirement by exposing a fragment that we'll be able to use in conjunction with `useFragment` to unmask the data from the query.

You have two solutions here, either you embrace Fragment Masking (see [Unleash the power of Fragments with GraphQL Codegen](https://the-guild.dev/blog/unleash-the-power-of-fragments-with-graphql-codegen)) or you disable Fragment Masking.

Thank you for the repo, it was useful! I sent some pull requests to clarify the possible solutions here.

### PR 1: Initial Setup

First pull request is to setup `client-preset` properly by leveraging generated `graphql` functions rather than `graphql.macro` (no need to type queries manually!).

🔗 Pull request: https://github.com/SimonPringleWallace/code-gen-issue/pull/1

### Solution 1: Embrace Fragment Masking

I highly recommend embracing fragment masking as I believe the data masking here is super important and will help you build better—and isolated—components.

You can read more about fragment masking here: 📚 [Unleash the power of Fragments with GraphQL Codegen](https://the-guild.dev/blog/unleash-the-power-of-fragments-with-graphql-codegen).

🔗 Pull request: https://github.com/SimonPringleWallace/code-gen-issue/pull/2

```typescript
const CharactersCountFragment = graphql(`
  fragment CharactersCount on Characters {
    info {
      count
    }
  }
`);

type CharactersCountProps = {
  characters: FragmentType<typeof CharactersCountFragment>;
};

const CharactersCount = (props: CharactersCountProps) => {
  const characters = useFragment(CharactersCountFragment, props.characters);

  return <div>{characters?.info?.count}</div>;
};
```

### Solution 2: Disable Fragment Masking

If you don't want to embrace fragment masking, we can disable them as described in the documentation: [How to disable Fragment Masking](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#how-to-disable-fragment-masking).

🔗 Pull request: https://github.com/SimonPringleWallace/code-gen-issue/pull/3

```diff
  ...
  generates: {
    "src/common/graphql/types/": {
      preset: "client",
+     presetConfig: {
+       fragmentMasking: false,
+     },
      plugins: [],
    },
  },
  ...
```

**Comment by SimonPringleWallace** [comment]
Thank you @charpeni! So appreciate the detailed explanation and especially the PR with comments. Extremely helpful. Thanks for all your work on this tool - it's already been a complete game changer for our dev team.
